File: android/app/src/main/java/com/almatraders/erp/shell/AlmaSession.kt
//
//  AlmaSession.kt
//  ALMA ERP — one shared identity/role source for the native screens (security audit
//  2026-07-12). Fetched once from GET /api/users/me (→ user.role, an AlmaRole string:
//  SUPER_ADMIN | ADMIN | HR | STAFF | VIEWER) + GET /api/assistant/more-pulse (isOwner,
//  businessAccess). Screens read the derived flags to HIDE privileged write UI for
//  non-admins — defense-in-depth so a missing server route-gate can't be reached
//  natively. The server stays the real authority; this only stops the app from
//  OFFERING an action the user can't perform.
//

package com.almatraders.erp.shell

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

object AlmaSession {

    /** AlmaRole string, or null until loaded / on failure (treat null as least-privilege). */
    var role by mutableStateOf<String?>(null)
        private set

    /** System owner (isSystemOwner on the server). */
    var isOwner by mutableStateOf(false)
        private set

    /** Our ERP User.id — the OneSignal external_id every push is addressed to.
     *  Without it OneSignal has no identity to attach this device to. */
    var userId by mutableStateOf<String?>(null)
        private set

    /** Business ids the user may access (ERP businessAccess). */
    var businessAccess by mutableStateOf(listOf<String>())
        private set

    @Volatile private var loaded = false

    /** App context (set once by the shell) so the session can drive NativePush.login. */
    @Volatile private var appContext: android.content.Context? = null
    fun attach(context: android.content.Context) {
        if (appContext == null) appContext = context.applicationContext
    }

    /** Bumps after every COMPLETED load. Screens key their fetch on this so a fresh
     *  login (which force-reloads the session) makes them re-fetch automatically —
     *  the fix for "role-based nav/data didn't appear after signing in". */
    var authVersion by mutableStateOf(0)
        private set

    /** Tri-state auth signal for the startup gate:
     *   null  = not yet checked / couldn't reach the server (offline) → don't gate,
     *   true  = /users/me returned a user → signed in,
     *   false = server said 401/403 → NOT signed in → show the native login gate.
     *  Only a definite `false` gates, so an offline launch with a valid cookie still
     *  opens the app instead of wrongly forcing a login. */
    var authed by mutableStateOf<Boolean?>(null)
        private set

    /** SUPER_ADMIN or ADMIN — the web `isAdminRole` gate for business-module writes. */
    val isAdmin: Boolean get() = isOwner || role == "SUPER_ADMIN" || role == "ADMIN"

    /** Trading/Digital admin-only writes (accounts, targets, HR, invoices, payments). */
    val canManageBusiness: Boolean get() = isAdmin

    /**
     * Normalised role used for nav gating. The system owner is always SUPER_ADMIN.
     * A not-yet-loaded / blank role falls back to STAFF (least-privilege) so the app
     * NEVER flashes privileged nav before the real role arrives — it only ever GAINS
     * sections once loaded, never wrongly shows them first.
     */
    val effectiveRole: String
        get() = when {
            isOwner -> AlmaRoles.SUPER_ADMIN
            role != null -> AlmaRoles.normalize(role)
            else -> AlmaRoles.STAFF
        }

    /** Should this role see/reach a route? Defense-in-depth mirror of the web nav gate. */
    fun canSee(path: String, businessId: String = AlmaBusinesses.LIFESTYLE): Boolean =
        AlmaAccess.isPathAllowedForRole(path, effectiveRole, businessId)

    suspend fun load(force: Boolean = false) {
        if (loaded && !force) return
        // Role from /api/users/me (authoritative role string). A 401/403 here is the
        // definite "not signed in" signal that drives the startup login gate; a network
        // error leaves `authed` unknown (null) so we never gate an offline-but-valid user.
        try {
            val me = AlmaApi.getObject("/api/users/me")
            val u = me.optJSONObject("user") ?: me.optJSONObject("data") ?: me
            u.str("id")?.let { userId = it }
            u.str("role")?.let { role = it }
            u.flexBool("isSystemOwner")?.let { if (it) isOwner = true }
            authed = true
        } catch (e: AlmaApiException.NotAuthenticated) {
            authed = false
        } catch (_: Exception) { /* offline / transient — leave authed unknown */ }
        // Owner flag + business access from more-pulse (best-effort).
        try {
            val pulse = AlmaApi.getObject("/api/assistant/more-pulse")
            val p = pulse.optJSONObject("data") ?: pulse
            p.flexBool("isOwner")?.let { if (it) isOwner = true }
            p.optJSONArray("businessAccess")?.let { arr ->
                businessAccess = (0 until arr.length()).mapNotNull { arr.optString(it).takeIf { s -> s.isNotEmpty() } }
            }
        } catch (_: Exception) { }
        // Attach this device to our user in OneSignal the moment we know who we are.
        // Push is addressed to this external_id; without the login the device is
        // anonymous and every targeted push (calls included) misses it.
        appContext?.let { ctx -> userId?.let { NativePush.login(ctx, it) } }
        loaded = true
        authVersion++          // signal every screen keyed on authVersion to re-fetch
    }

    /** Force a fresh identity fetch — call right after a successful sign-in so the role
     *  (and owner flag) update from the fail-closed STAFF default to the real values,
     *  and every screen keyed on [authVersion] reloads. Resets owner first so an
     *  account switch can't keep a stale elevated flag. */
    suspend fun reload() {
        isOwner = false
        role = null
        loaded = false
        load(force = true)
    }

    /** Clear all identity on explicit sign-out (fail closed to STAFF + gate to login). */
    fun signedOut() {
        isOwner = false
        role = null
        businessAccess = emptyList()
        loaded = false
        authed = false
        authVersion++
    }
}

/** Ensures AlmaSession is loaded once when a gated screen appears. Read the flags after. */
@Composable
fun RememberSession() {
    LaunchedEffect(Unit) { AlmaSession.load() }
}

File: ios/App/App/AlmaSession.swift
//
//  AlmaSession.swift
//  ALMA ERP — ONE shared identity for the native shell (role · owner flag ·
//  business access · current business). Mirrors the web's ActorContext +
//  BusinessContext: every tab, More row, deep link and write button asks THIS
//  object, never its own ad-hoc `/api/users/me` fetch.
//
//  Source of truth: GET /api/users/me → { user: { id, name, role, isSystemOwner,
//  businessAccess } } (the same route the web's drawer trusts). Loaded at launch,
//  re-loaded after a native sign-in and on foreground, cleared on sign-out.
//
//  Fail CLOSED: until the server answers, `effectiveRole` is VIEWER (the web's own
//  normalisation of an unknown role) and the shell offers only what a VIEWER may
//  see — nothing privileged ever flashes. To avoid a visible rebuild on every cold
//  start, the LAST CONFIRMED identity is cached per backend host and used
//  instantly; a server answer that differs triggers one rebuild. A 401 keeps the
//  cache (transient on cold start) — only an explicit sign-out wipes it.
//
//  Current business mirrors the web rule (BusinessContext): the stored choice if
//  the user may access it, else the first allowed business. A Trading-only staff
//  member therefore lands in Trading, never on a Lifestyle screen that 403s.
//

import Foundation
import Observation

extension Notification.Name {
    /// Posted on the main thread when role / owner / business access / user id
    /// CHANGED after a load — the tab bar rebuilds itself from the new identity.
    static let almaSessionChanged = Notification.Name("almaSessionChanged")
    /// Posted when the current business changed (switcher or access clamp).
    static let almaBusinessChanged = Notification.Name("almaBusinessChanged")
}

@available(iOS 17.0, *)
@Observable
@MainActor
final class AlmaSession {

    static let shared = AlmaSession()

    // MARK: Identity

    /// Server role string normalised, or nil until the first load completes.
    private(set) var role: AlmaRole?
    private(set) var isOwner = false
    private(set) var userId: String?
    private(set) var name: String = ""
    /// Businesses the user may switch to. Empty until loaded (→ treated as none).
    private(set) var businessAccess: [AlmaBusinessId] = []
    /// Tri-state auth: nil = not yet known / offline, true = signed in, false = 401.
    private(set) var authed: Bool?
    /// Bumps after every COMPLETED load (screens key their refetch on it).
    private(set) var authVersion = 0
    private(set) var loaded = false

    /// The role every gate uses. Owner ⇒ SUPER_ADMIN; otherwise the loaded role;
    /// not loaded ⇒ VIEWER (least privilege, same as the web's normalizer).
    var effectiveRole: AlmaRole {
        if isOwner { return .SUPER_ADMIN }
        return role ?? .VIEWER
    }

    /// Businesses the user may access — the owner always has all three
    /// (`normalizeBusinessAccessForRole`).
    var allowedBusinesses: [AlmaBusinessId] {
        AlmaBusinessId.accessForRole(businessAccess, role: effectiveRole)
    }

    var isAdmin: Bool { effectiveRole == .SUPER_ADMIN || effectiveRole == .ADMIN }

    // MARK: Current business

    private static let businessKey = "alma-business-id"   // web STORAGE_KEY

    private var storedBusiness: AlmaBusinessId? {
        AlmaBusinessId(rawValue: UserDefaults.standard.string(forKey: Self.businessKey) ?? "")
    }

    /// The business the shell is in — BusinessContext's rule: stored if allowed,
    /// else the first allowed, else Lifestyle (before access is known).
    var businessId: AlmaBusinessId {
        let allowed = allowedBusinesses
        if let stored = storedBusiness, allowed.isEmpty || allowed.contains(stored) { return stored }
        return allowed.first ?? AlmaBusinessId.default
    }

    /// Switch business (the More "Business" pill). Ignored when not allowed.
    func setBusiness(_ id: AlmaBusinessId) {
        guard allowedBusinesses.contains(id) else { return }
        let before = businessId
        UserDefaults.standard.set(id.rawValue, forKey: Self.businessKey)
        if before != businessId {
            NotificationCenter.default.post(name: .almaBusinessChanged, object: nil)
        }
    }

    // MARK: Gates

    /// Can this user SEE / open `path`? The web's two gates combined: the route must
    /// be valid for the business it belongs to (proxy.ts + BusinessContext) AND
    /// the user must have access to that business.
    func canSee(_ path: String) -> Bool {
        let bare = path.split(separator: "?").first.map(String.init) ?? path
        let business = AlmaAccess.business(for: bare, current: businessId)
        // Login / recovery / public share pages never need a business.
        if bare.hasPrefix("/login") || bare.hasPrefix("/forgot-password")
            || bare.hasPrefix("/reset-password") || bare.hasPrefix("/invoice/share") { return true }
        guard allowedBusinesses.contains(business) else { return false }
        return AlmaAccess.isPathAllowedForRole(bare, role: effectiveRole, business: business)
    }

    /// `can(role, capability)` for the signed-in user.
    func can(_ capability: AlmaCapability) -> Bool {
        AlmaAccess.can(effectiveRole, capability)
    }

    /// The web nav this user sees in the current business.
    var nav: [AlmaNavItem] {
        AlmaAccess.nav(role: effectiveRole, business: businessId)
    }

    /// Where the web lands this user (`roleHomePath`).
    var homePath: String {
        AlmaAccess.roleHomePath(role: effectiveRole, business: businessId)
    }

    // MARK: Loading

    private struct MeResponse: Decodable {
        struct User: Decodable {
            let id: String?
            let name: String?
            let role: String?
            let isSystemOwner: Bool?
            let businessAccess: String?
        }
        let user: User?
    }

    /// Identity tuple that decides the tab bar — a change rebuilds the shell.
    private struct Snapshot: Codable, Equatable {
        var userId: String?
        var role: AlmaRole?
        var isOwner: Bool
        var businessAccess: [AlmaBusinessId]
        var name: String
    }

    private var snapshot: Snapshot {
        Snapshot(userId: userId, role: role, isOwner: isOwner, businessAccess: businessAccess, name: name)
    }

    private static var cacheKey: String { "alma.session.cache." + (AlmaAPI.baseURL.host ?? "prod") }

    private var inflight: Task<Void, Never>?

    private init() {
        restoreCache()
    }

    /// Seed from the last confirmed identity so the first tab bar is already right
    /// for a returning user (no VIEWER → real-role flicker). Nothing is trusted
    /// beyond what the server confirmed last time on this same host.
    private func restoreCache() {
        guard let data = UserDefaults.standard.data(forKey: Self.cacheKey),
              let s = try? JSONDecoder().decode(Snapshot.self, from: data) else { return }
        apply(s)
    }

    private func apply(_ s: Snapshot) {
        userId = s.userId
        role = s.role
        isOwner = s.isOwner
        businessAccess = s.businessAccess
        name = s.name
    }

    private func persist(_ s: Snapshot) {
        if let data = try? JSONEncoder().encode(s) {
            UserDefaults.standard.set(data, forKey: Self.cacheKey)
        }
    }

    /// Load (or re-load) identity. Coalesces concurrent callers.
    func load(force: Bool = false) async {
        #if DEBUG
        if fixtureLocked { return }
        #endif
        if loaded && !force { return }
        if let inflight { await inflight.value; return }
        let task = Task { await self.fetch() }
        inflight = task
        await task.value
        inflight = nil
    }

    /// Force a fresh fetch — right after a native sign-in, or on foreground.
    func reload() async { await load(force: true) }

    private var lastLoadedAt: Date?

    /// Foreground re-confirmation, at most once a minute.
    func reloadIfStale(_ interval: TimeInterval = 60) async {
        if let last = lastLoadedAt, Date().timeIntervalSince(last) < interval { return }
        await load(force: true)
    }

    private func fetch() async {
        let before = snapshot
        let beforeBusiness = businessId
        do {
            var me: MeResponse
            do {
                me = try await AlmaAPI.shared.get("/api/users/me")
            } catch AlmaAPIError.http(let status, _) where status == 403 {
                // /users/me defaults to Lifestyle when the user has several
                // businesses — a CDIT+Trading user 403s there. Ask again for a
                // business we know (or each in turn) before giving up.
                me = try await fetchMeAnyBusiness()
            }
            guard let u = me.user else { return }
            let next = Snapshot(
                userId: u.id,
                role: AlmaRole.normalize(u.role),
                isOwner: u.isSystemOwner ?? false,
                businessAccess: AlmaBusinessId.parseAccess(u.businessAccess),
                name: (u.name ?? "").trimmingCharacters(in: .whitespacesAndNewlines))
            apply(next)
            persist(next)
            authed = true
        } catch AlmaAPIError.notAuthenticated {
            authed = false
        } catch {
            // Offline / transient: keep whatever we had; authed stays unknown.
        }
        loaded = true
        lastLoadedAt = Date()
        authVersion += 1
        if snapshot != before {
            NotificationCenter.default.post(name: .almaSessionChanged, object: nil)
        }
        if businessId != beforeBusiness {
            NotificationCenter.default.post(name: .almaBusinessChanged, object: nil)
        }
    }

    private func fetchMeAnyBusiness() async throws -> MeResponse {
        var candidates = businessAccess
        for b in AlmaBusinessId.all where !candidates.contains(b) { candidates.append(b) }
        var lastError: Error = AlmaAPIError.http(status: 403, body: "")
        for b in candidates {
            do {
                return try await AlmaAPI.shared.get("/api/users/me", query: ["business_id": b.rawValue])
            } catch AlmaAPIError.http(let status, _) where status == 403 {
                lastError = AlmaAPIError.http(status: status, body: "")
                continue
            }
        }
        throw lastError
    }

    /// Explicit sign-out: wipe identity + cache, fail closed, rebuild the shell.
    func signedOut() {
        let before = snapshot
        userId = nil; role = nil; isOwner = false; businessAccess = []; name = ""
        authed = false
        loaded = false
        UserDefaults.standard.removeObject(forKey: Self.cacheKey)
        authVersion += 1
        if snapshot != before {
            NotificationCenter.default.post(name: .almaSessionChanged, object: nil)
        }
    }

    #if DEBUG
    /// Headless sim self-test hook (DEBUG only, never ships): `ALMA_ACCESS_FIXTURE=
    /// ROLE[:BUSINESS_ACCESS_CSV]` pins the identity so every role × business tab
    /// bar / More menu can be screenshot-verified without five real sign-ins.
    /// Example: SIMCTL_CHILD_ALMA_ACCESS_FIXTURE=STAFF:ALMA_TRADING
    func applyFixtureIfRequested() {
        guard let raw = ProcessInfo.processInfo.environment["ALMA_ACCESS_FIXTURE"], !raw.isEmpty else { return }
        let parts = raw.split(separator: ":", maxSplits: 1).map(String.init)
        let r = AlmaRole.normalize(parts.first)
        let access = parts.count > 1 ? AlmaBusinessId.parseAccess(parts[1]) : AlmaBusinessId.all
        apply(Snapshot(userId: "fixture", role: r, isOwner: r == .SUPER_ADMIN,
                       businessAccess: access, name: "Fixture \(r.label)"))
        authed = true
        loaded = true
        fixtureLocked = true
        authVersion += 1
        NotificationCenter.default.post(name: .almaSessionChanged, object: nil)
    }
    /// While a fixture is pinned, server loads must not overwrite it.
    private(set) var fixtureLocked = false
    #endif
}

#if DEBUG
@available(iOS 17.0, *)
extension AlmaSession {
    /// `load` honours a pinned fixture (DEBUG only).
    var isFixtureLocked: Bool { fixtureLocked }
}
#endif
